From 4377f5de702a1621271cb0097316dc4b3e7f660e Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Fri, 9 Jul 2021 17:15:50 +0200
Subject: [PATCH 2/3] drbd_bitmap: Do not use __free_pages: use put_page() to
 honor reference count.

Also use alloc_page() for allocating the page in bm_page_io_async()
alloc_page initializes more struct page members.
---
 drbd/drbd_bitmap.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drbd/drbd_bitmap.c b/drbd/drbd_bitmap.c
index 893426a..cbaf9f5 100644
--- a/drbd/drbd_bitmap.c
+++ b/drbd/drbd_bitmap.c
@@ -335,7 +335,7 @@ static void bm_free_pages(struct page **pages, ULONG_PTR number)
 				 i, number);
 			continue;
 		}
-		__free_page(pages[i]);
+		put_page(pages[i]);
 		pages[i] = NULL;
 	}
 }
@@ -1151,8 +1151,11 @@ static void bm_page_io_async(struct drbd_bm_aio_ctx *ctx, int page_nr)
 	bm_set_page_unchanged(b->bm_pages[page_nr]);
 
 	if (ctx->flags & BM_AIO_COPY_PAGES) {
+/*
 		page = mempool_alloc(&drbd_md_io_page_pool,
 				GFP_NOIO | __GFP_HIGHMEM);
+*/
+		page = alloc_page(GFP_NOIO | __GFP_HIGHMEM);
 		expect(device, page);
 
 		page->private = b->bm_pages[page_nr]->private;
-- 
2.17.0

